### PR TITLE
rangefeed: fix handling of aborted transactions found by txnPushAttempt

### LINKED ISSUE #35816: roachtest: cdc/crdb-chaos/rangefeed=true failed
SHA: https://github.com/cockroachdb/cockroach/commits/04ef15974085e14f758b20c552a84052eac9fa2b

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=cdc/crdb-chaos/rangefeed=true PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1180753&tab=buildLog

```
The test failed on master:
	cluster.go:1267,cdc.go:625,cdc.go:125,cluster.go:1605,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1180753-cdc-crdb-chaos-rangefeed-true:4 -- ./workload run tpcc --warehouses=100 --duration=30m --tolerate-errors {pgurl:1-3}  returned:
		stderr:
		
		stdout:
		    1.8      0.0      0.0      0.0      0.0 delivery
		   3m21s      585            0.0           17.4      0.0      0.0      0.0      0.0 newOrder
		   3m21s      585            0.0            1.7      0.0      0.0      0.0      0.0 orderStatus
		   3m21s      585            0.0           18.9      0.0      0.0      0.0      0.0 payment
		   3m21s      585            0.0            2.0      0.0      0.0      0.0      0.0 stockLevel
		E190316 06:21:18.754959 1 workload/cli/run.go:420  error in payment: dial tcp 10.142.0.129:26257: connect: connection refused
		   3m22s      607            0.0            1.8      0.0      0.0      0.0      0.0 delivery
		   3m22s      607            0.0           17.3      0.0      0.0      0.0      0.0 newOrder
		   3m22s      607            0.0            1.7      0.0      0.0      0.0      0.0 orderStatus
		   3m22s      607            0.0           18.8      0.0      0.0      0.0      0.0 payment
		   3m22s      607            0.0            2.0      0.0      0.0      0.0      0.0 stockLevel
		: signal: killed
	cluster.go:1626,cdc.go:213,cdc.go:433,test.go:1214: unexpected node event: 2: dead

```

### PR DESCRIPTION
**Pull request by nvb**
Fixes #35816.

The fix in #35777 inadvertently broke `txnPushAttempt` and its interaction with
transactions that were found to be aborted. This commit restores this handling
by reverting part of the assertion added in #35772 and splitting up the handling
of "aborted intents" from the handling of "aborted transactions".

Aborted transactions were always one of the most tricky situations to handle in
rangefeed. The challenge is that rangefeed may be tracking a transaction that
has been abandoned and may want to push it to unblock its resolved timestamp,
but it doesn't have a way to know where the aborted transaction's intents live
on its range. The aborted transaction record can't tell us this either. So the
only thing rangefeed can do is push the transaction and convince itself that the
aborted intents on its range can be safely ignored. This poses a problem for
accurate reference counting of intents though. We need to be able to discard a
transaction and still handle the eventual resolution of its aborted intents.

To handle this properly, the change splits up `MVCCAbortIntentOp` into
`MVCCAbortIntentOp` and `MVCCAbortTxnOp`. Before this change, the single
operation type was attempting to fill both roles, but they mean fundamentally
different things. `MVCCAbortTxnOp` is then given back its ability to delete
transactions from the queue directly without worrying about their reference
count. Next, the `unresolvedIntentQueue` is extended to be a little more
flexible. Instead of asserting that reference counts never drop below zero, it just
ignores changes that would make reference counts negative. This has the potential
to "leak" reference counts that saturate at zero and are then incremented again,
but we know that this will only ever happen for aborted transactions, which can
always be pushed again to rediscover that they have been aborted.

Release note: None

### CODE REVIEW

**Review by danhhz — COMMENTED** [review]
@tbg and/or @bdarnell can I get your eyes on this, too? I'm not quite the careful reviewer that the two of you are and it seems like this is all more complicated than anyone initially thought

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35889#-:-LaH9ZIc6Es9I7smguwc:b-pqcytl)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @danhhz and @nvanbenschoten)

___
*[pkg/storage/engine/enginepb/mvcc3.proto, line 247 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35889#-LaH93pG8asNtlOuP_YD:-LaH93pG8asNtlOuP_YE:b-hhqv9m) ([raw file](https://github.com/cockroachdb/cockroach/blob/9628a321e4736550b051c5c11ed9ff18075484fe/pkg/storage/engine/enginepb/mvcc3.proto#L247)):*
> ```protobuf
>   MVCCCommitIntentOp commit_intent = 4;
>   MVCCAbortIntentOp  abort_intent  = 5;
>   MVCCAbortTxnOp     abort_txn     = 6;
> ```

What happens if one of these gets sent during a rolling upgrade to a node that doesn't have this branch in the `if`? Looks like a panic?

**Review by nvb — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35889)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell, @danhhz, and @tbg)

___
*[pkg/storage/engine/enginepb/mvcc3.proto, line 247 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35889#-LaH93pG8asNtlOuP_YD:-LaHBBW0ANczNI_1G7hx:bpfa4ue) ([raw file](https://github.com/cockroachdb/cockroach/blob/9628a321e4736550b051c5c11ed9ff18075484fe/pkg/storage/engine/enginepb/mvcc3.proto#L247)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

What happens if one of these gets sent during a rolling upgrade to a node that doesn't have this branch in the `if`? Looks like a panic?
</blockquote></details>

We only ever send them to our local rangefeed from `txnPushAttempt` tasks, so that's not a concern.

File: pkg/storage/rangefeed/resolved_timestamp.go [hunk 1]
```diff
@@ -414,6 +462,24 @@ func (uiq *unresolvedIntentQueue) updateTxn(
 	return false
 }
 
+// Del removes the transaction from the queue. It returns whether the update had
```
Comment (nvanbenschoten) [inline pkg/storage/rangefeed/resolved_timestamp.go]: This method is exactly what was removed in https://github.com/cockroachdb/cockroach/pull/35777/files#diff-425bf0ccd6243bd7bb88910fc308bdf4L424.

**Review by tbg — COMMENTED** [review]
What's the reasoning behind clamping the refcounts to zero? If we let them go negative, wouldn't they either go back to zero or the txn would be deleted wholesale thanks to having been aborted? I'm asking because a) that seems conceptually simpler and b) perhaps we could retain some assertions? For example, if the queue is init'ed and we're seeing a negative refcount, then either it goes away organically or we eventually find that the txn is aborted. In particular, a "stuck" negative refcount for which we discover a committed txn would trigger an error. Maybe this is is also possible with the clamping, not sure. I don't understand the significance of the value zero. I thought once a refcount of zero was reached, the txn would be deleted from the queue, but I don't think that's happening here.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35889#-:-LaKgYSm1rT9E7Xd346g:bcyz824)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell, @nvanbenschoten, and @tbg)

___
*[pkg/storage/engine/enginepb/mvcc3.proto, line 219 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35889#-LaKeR1971ZZsQlf1iDv:-LaKeR1971ZZsQlf1iDw:bz45ybg) ([raw file](https://github.com/cockroachdb/cockroach/blob/9628a321e4736550b051c5c11ed9ff18075484fe/pkg/storage/engine/enginepb/mvcc3.proto#L219)):*
> instance
___
*[pkg/storage/rangefeed/processor_test.go, line 667 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35889#-LaKekH0B5lWex-109M7:-LaKekH0B5lWex-109M8:b-wx4dqt) ([raw file](https://github.com/cockroachdb/cockroach/blob/9628a321e4736550b051c5c11ed9ff18075484fe/pkg/storage/rangefeed/processor_test.go#L667)):*
> ```Go
> 	p.ConsumeLogicalOps(
> 		writeIntentOpWithKey(txn1Meta.ID, txn1Meta.Key, txn1Meta.Timestamp),
> 		writeIntentOpWithKey(txn2Meta.ID, txn2Meta.Key, txn2Meta.Timestamp),
> ```

What's this change about?

**Review by nvb — COMMENTED** [review]
> What's the reasoning behind clamping the refcounts to zero?
> I thought once a refcount of zero was reached, the txn would be deleted from the queue

That is what's happening here. We never store ref counts of zero in the queue. We always remove it in that case. You can think of a `MVCCAbortIntentOp` as causing a `-1` refcount delta and a `MVCCAbortTxnOp` as a causing a `-count` refcount delta. But notice that we're now overcounting intents, so we need to clamp at zero. This is ok though, because we know that we'll only ever be saturating at zero for aborted transactions.

I think you might be suggesting that we let ref counts go negative when we overcount and then rely on a second `MVCCAbortTxnOp` to bring the count back to zero. That wouldn't work because it could result in a txn being added back into the queue with a count of -1 after a `MVCCAbortIntentOp` and causing a resolved timestamp regression.

Did that answer your question?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35889#-:-LaKkU7h4WYDeDLe1vLJ:b64ltgq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell and @tbg)

___
*[pkg/storage/engine/enginepb/mvcc3.proto, line 219 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35889#-LaKeR1971ZZsQlf1iDv:-LaKk1Qe7MmP-QL44XCA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9628a321e4736550b051c5c11ed9ff18075484fe/pkg/storage/engine/enginepb/mvcc3.proto#L219)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

> instance
</blockquote></details>

Done.
___
*[pkg/storage/rangefeed/processor_test.go, line 667 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35889#-LaKekH0B5lWex-109M7:-LaKk2aT2vriNa4nlp9D:b-4b97jt) ([raw file](https://github.com/cockroachdb/cockroach/blob/9628a321e4736550b051c5c11ed9ff18075484fe/pkg/storage/rangefeed/processor_test.go#L667)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

What's this change about?
</blockquote></details>

This test was erroneously passing before this change because it only had a single intent for the transaction that was eventually discovered as aborted. This allowed the single ref count decrement to function as a full deletion. With this change, the test now only passes when an aborted txn results in a `MVCCAbortTxnOp` instead of a `MVCCAbortIntentOp`.

**Comment by nvanbenschoten** [comment]
Also for what it's worth, this is how this all worked up until #35777. The logic here isn't really new, it's just being adapted to work with that PR, which realized that aborted intents don't always indicate aborted transactions.

**Review by tbg — COMMENTED** [review]
Yep, thanks

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35889#-:-LaL-cmkBr3Fn2mfP5n9:b-c3iuc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell, @nvanbenschoten, and @tbg)

___
*[pkg/storage/rangefeed/resolved_timestamp.go, line 454 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35889#-LaKyAt-2nRG69eesLIV:-LaKyAt0Bz_2R6xFFD2z:bsakh6i) ([raw file](https://github.com/cockroachdb/cockroach/blob/9628a321e4736550b051c5c11ed9ff18075484fe/pkg/storage/rangefeed/resolved_timestamp.go#L454)):*
> ```Go
> 	} else if txn.refCount < 0 && !uiq.allowNegRefCount {
> 		// Ignore txn.
> 		return false
> ```

Already taking into account that the callers only use deltas of 0, 1, -1 (I had to go look that up because it's not obvious, so perhaps the code should be written more robustly, with larger negative deltas it seems that more oddities come up incl negative final refcounts) so it looks like would only be able to hit the branch if `txn.refCount` before the `delta` was `0` but then how is it in the queue? The way I picture this is

- txn is aborted and deleted from queue
- txn writes intent a, gets a refcount of one
- txn intent gets aborted, refcount -1 hits the deletion branch

or

- txn aborted and deleted
- txn intent gets aborted (hits the !ok branch above)

so what example hits this code? (might want to add it as a comment while you're at it)

**Review by tbg — COMMENTED** [review]
> Reviewed 9 of 11 files at r1, 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35889)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell and @nvanbenschoten)

**Review by nvb — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35889)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell and @tbg)

___
*[pkg/storage/rangefeed/resolved_timestamp.go, line 454 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35889#-LaKyAt-2nRG69eesLIV:-LaLd4IUBYPD6TBHOePu:b-aykbt5) ([raw file](https://github.com/cockroachdb/cockroach/blob/9628a321e4736550b051c5c11ed9ff18075484fe/pkg/storage/rangefeed/resolved_timestamp.go#L454)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Already taking into account that the callers only use deltas of 0, 1, -1 (I had to go look that up because it's not obvious, so perhaps the code should be written more robustly, with larger negative deltas it seems that more oddities come up incl negative final refcounts) so it looks like would only be able to hit the branch if `txn.refCount` before the `delta` was `0` but then how is it in the queue? The way I picture this is

- txn is aborted and deleted from queue
- txn writes intent a, gets a refcount of one
- txn intent gets aborted, refcount -1 hits the deletion branch

or

- txn aborted and deleted
- txn intent gets aborted (hits the !ok branch above)

so what example hits this code? (might want to add it as a comment while you're at it)
</blockquote></details>

Yeah this case wasn't being exercised by the external interface of the type, but I included it for robustness. You're completely right though that it's not actually what we want. Instead, we want it to be collapsed in with the previous case like:

```
if txn.refCount == 0 || (txn.refCount < 0 && !uiq.allowNegRefCount) {
```
___
*[pkg/storage/rangefeed/resolved_timestamp.go, line 467 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35889#-LaLdf8pBLlBQe97wzgj:-LaLdf8pBLlBQe97wzgk:b28toe8) ([raw file](https://github.com/cockroachdb/cockroach/blob/145d1564ae12b9e4070f2efd87e341703f57e065/pkg/storage/rangefeed/resolved_timestamp.go#L467)):*
> ```Go
> // Del removes the transaction from the queue. It returns whether the update had
> // an effect on the oldest transaction in the queue.
> func (uiq *unresolvedIntentQueue) Del(txnID uuid.UUID) bool {
> ```

@tbg I asked on slack, but I'll keep the conversation on here. Would you find the following more clear?
```
func (uiq *unresolvedIntentQueue) Del(txnID uuid.UUID) bool {
	if uiq.allowNegRefCount {
		panic("cannot Del while allowNegRefCount")
	}
	return uiq.updateTxn(txnID, nil, hlc.Timestamp{}, math.MinInt64)
}
```

To avoid underflow, we'd need the `allowNegRefCount` check, which is kind of gross.

**Review by nvb — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35889)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell and @tbg)

___
*[pkg/storage/rangefeed/resolved_timestamp.go, line 454 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35889#-LaKyAt-2nRG69eesLIV:-LaLtb5Q2dXdiO6qGp1P:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9628a321e4736550b051c5c11ed9ff18075484fe/pkg/storage/rangefeed/resolved_timestamp.go#L454)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Yeah this case wasn't being exercised by the external interface of the type, but I included it for robustness. You're completely right though that it's not actually what we want. Instead, we want it to be collapsed in with the previous case like:

```
if txn.refCount == 0 || (txn.refCount < 0 && !uiq.allowNegRefCount) {
```
</blockquote></details>

Done.
___
*[pkg/storage/rangefeed/resolved_timestamp.go, line 467 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35889#-LaLdf8pBLlBQe97wzgj:-LaLtZcjB3prH1NiAwsA:bhip6p0) ([raw file](https://github.com/cockroachdb/cockroach/blob/145d1564ae12b9e4070f2efd87e341703f57e065/pkg/storage/rangefeed/resolved_timestamp.go#L467)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

@tbg I asked on slack, but I'll keep the conversation on here. Would you find the following more clear?
```
func (uiq *unresolvedIntentQueue) Del(txnID uuid.UUID) bool {
	if uiq.allowNegRefCount {
		panic("cannot Del while allowNegRefCount")
	}
	return uiq.updateTxn(txnID, nil, hlc.Timestamp{}, math.MinInt64)
}
```

To avoid underflow, we'd need the `allowNegRefCount` check, which is kind of gross.
</blockquote></details>

Ok, I added this as a comment instead.

**Comment by nvanbenschoten** [comment]
@tbg I'm not sure whether you were planning on giving this another look or not. I'm hoping to get it in for tonight's roachtests.

**Review by bdarnell — APPROVED** [review]
LGTM, although I'm not very familiar with this code. 

> Reviewed 8 of 11 files at r1, 2 of 2 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35889#-:-LaMbjmrBMzgpCSDDZVt:b-78s6gt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

**Comment by nvanbenschoten** [comment]
TFTRs!

bors r+

**Comment by tbg** [comment]
Merge it, I'll take another pass tomorrow but have no reason to assume
there's anything I'd want to have changed

**Review by tbg — APPROVED** [review]
> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35889)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Pull request merged**